packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +424,411 @@ contract AmmPool is IBlockReceiver {
+                        require(transfer.storageID == join.storageIDs[i], "INVALID_TX_DATA");
+                    }
+                    if (signature.length != 0) {
+                        // Now approve this transfer
+                        transfer.validUntil = 0xffffffff;
+                        bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
+                        exchange.approveTransaction(join.owner, txHash);
+                    }
+                    ctx.numTransactionsConsumed++;
+                    // Update the amount to the actual amount transferred (which can have some some small rounding errors)
+                    amount = transfer.amount;
+                    // Update the balances in the account
+                    ctx.ammBalancesInAccount[i] = ctx.ammBalancesInAccount[i].add(amount);
+                } else {
                     // Make the amount unavailable for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].sub(amount);
-                }
-            } else if (item.workType == WorkType.EXIT) {
-                PoolExit memory exit = item.exit;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = exit.minAmountsOut[i];
-                    ammBalances[i] = ammBalances[i].sub(amount);
-                    // Make the amount available for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].add(amount);
+                    address token = ctx.tokens[i].addr;
+                    balance[token][join.owner] = balance[token][join.owner].sub(amount);
                 }
+                ctx.ammBalances[i] = ctx.ammBalances[i].add(amount);
             }
-            // Clean up the queue item to get the gas refund
-            delete queue[n];
+
+            // Mint liquidity tokens
+            mint(join.owner, join.poolAmountOut);
         }
-        queuePos += auxData.numItems;
 
-        // Deposit/Withdraw to/from the AMM account when necessary
-        for (uint i = 0; i < tokens.length; i++) {
-            if (ammBalances[i] > ammBalancesInAccount[i]) {
-                uint amount = ammBalances[i] - ammBalancesInAccount[i];
-                // Check that the deposit in the block matches the expected deposit
-                DepositTransaction.Deposit memory _deposit = _block.readDeposit(txIdx++);
-                require(_deposit.owner == address(this), "INVALID_TX_DATA");
-                require(_deposit.accountID == accountID, "INVALID_TX_DATA");
-                require(_deposit.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
-                require(_deposit.amount == amount, "INVALID_TX_DATA");
-                // Now do this deposit
-                uint ethValue = 0;
-                if (tokens[i].addr == address(0)) {
-                    ethValue = _deposit.amount;
+        if (!join.fromLayer2) {
```

Comment:
should we skip this if `valid` is false?

Reply:
Currently the funds can only be unlocked when processed by the queue and are locked specifically to process the work in the queue. So even if it's invalid we want to unlock the funds, otherwise they would remain locked.